pkg/gardenlet/controller/backupbucket/reconciler.go

```diff
@@ -0,0 +1,441 @@
+				Namespace: extensionSecret.Namespace,
+			},
+		}
+	)
+
+	if err := r.SeedClient.Get(ctx, client.ObjectKeyFromObject(extensionSecret), extensionSecret); err == nil {
+		if !reflect.DeepEqual(extensionSecret.Data, secret.Data) {
+			reconcileExtensionBackupBucket = true
+		}
+	}
+
+	if err := r.deployBackupBucketExtensionSecret(ctx, backupBucket); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err := r.SeedClient.Get(ctx, client.ObjectKeyFromObject(extensionBackupBucket), extensionBackupBucket); err != nil ||
+		!reflect.DeepEqual(extensionBackupBucket.Spec, expectedExtensionBackupBucketSpec) {
+		reconcileExtensionBackupBucket = true
+	} else {
+		var (
+			lastObservedError error
+			lastError         = extensionBackupBucket.Status.LastError
+			lastOperation     = extensionBackupBucket.Status.LastOperation
+		)
+
+		if lastError != nil || lastOperation == nil || lastOperation.State != gardencorev1beta1.LastOperationStateSucceeded {
+			reconcileExtensionBackupBucket = true
+			lastErrorPresent = true
+
+			if lastError != nil {
+				lastObservedError = v1beta1helper.NewErrorWithCodes(fmt.Errorf("error during reconciliation: %s", lastError.Description), lastError.Codes...)
+			} else if lastOperation == nil {
+				lastObservedError = fmt.Errorf("extension did not record a last operation yet")
+			} else {
+				lastObservedError = fmt.Errorf("extension state is not succeeded but %v", lastOperation.State)
+			}
+
+			lastObservedError = v1beta1helper.NewErrorWithCodes(lastObservedError, v1beta1helper.DeprecatedDetermineErrorCodes(lastObservedError)...)
+			reconcileErr := &gardencorev1beta1.LastError{
+				Codes:       v1beta1helper.ExtractErrorCodes(lastObservedError),
+				Description: lastObservedError.Error(),
+			}
+
+			r.Recorder.Eventf(backupBucket, corev1.EventTypeWarning, gardencorev1beta1.EventReconcileError, "%s", reconcileErr.Description)
+
+			if updateErr := updateBackupBucketStatusError(ctx, r.GardenClient, backupBucket, reconcileErr.Description+". Operation will be retried.", reconcileErr, r.Clock); updateErr != nil {
+				return reconcile.Result{}, fmt.Errorf("could not update status after reconciliation error: %w", updateErr)
+			}
+		} else {
+			var gardenGeneratedSecretRef *corev1.SecretReference
+
+			if extensionBackupBucket.Status.GeneratedSecretRef != nil {
+				seedGeneratedSecret, err := kutil.GetSecretByReference(ctx, r.SeedClient, extensionBackupBucket.Status.GeneratedSecretRef)
+				if err != nil {
+					return reconcile.Result{}, err
+				}
+
+				gardenGeneratedSecret := &corev1.Secret{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      generateGeneratedBackupBucketSecretName(backupBucket.Name),
+						Namespace: r.GardenNamespace,
+					},
+				}
+				ownerRef := metav1.NewControllerRef(backupBucket, gardencorev1beta1.SchemeGroupVersion.WithKind("BackupBucket"))
+
+				if _, err := controllerutils.CreateOrGetAndStrategicMergePatch(ctx, r.GardenClient, gardenGeneratedSecret, func() error {
```

Comment:
Why do we need strategic merge patch?

Reply:
I did not change this. Will use `controllerutils.GetAndCreateOrMergePatch`